translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,382 @@
+// limitations under the License.
+
+package opencensus
+
+import (
+	"strings"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+func ocToInternal(td consumerdata.TraceData) data.TraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := data.NewResourceSpans(resource, nil)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+	if spanCount == 0 {
+		return data.NewTraceData(resourceSpanList)
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	spans := data.NewSpanSlice(spanCount)
+	spanPtrs := make([]*data.Span, 0, spanCount)
+
+	for i, ocSpan := range td.Spans {
+		if ocSpan == nil {
+			// Skip nil spans.
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		destSpan := &spans[i]
+
+		ocSpanToInternal(destSpan, ocSpan)
+
+		if ocSpan.Resource != nil {
+			// Add a separate ResourceSpans item just for this span since it
+			// has a different Resource.
+			separateRS := data.NewResourceSpans(
+				ocNodeResourceToInternal(nil, ocSpan.Resource),
+				[]*data.Span{destSpan},
+			)
+			resourceSpanList = append(resourceSpanList, separateRS)
+		} else {
+			// Otherwise add the span to the first ResourceSpans item.
+			spanPtrs = append(spanPtrs, destSpan)
+		}
+	}
+
+	resourceSpans.SetSpans(spanPtrs)
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	dest.SetTraceID(data.TraceIDFromBytes(src.TraceId))
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(src.Name.GetValue())
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(internal.TimestampToUnixnano(src.StartTime))
+	dest.SetEndTime(internal.TimestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
```

Comment:
I am a bit confused that for data.Attributes we included the dropped property, but for events and links we didn't. TODO or issue is fine. Maybe we can create data.Events and data.Links to help with that.

Reply:
Yeah, it is probably worth adding additional helper types for symmetry.

Reply:
Added TODO.